x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java

```diff
@@ -182,11 +211,44 @@ public DownsampleIndexerAction.ShardDownsampleResponse execute() throws IOExcept
             );
         }
 
-        task.setRollupShardIndexerStatus(RollupShardIndexerStatus.COMPLETED);
+        task.updatePersistentTaskState(
+            new RollupShardPersistentTaskState(RollupShardIndexerStatus.COMPLETED, null),
+            ActionListener.wrap(response -> task.setRollupShardIndexerStatus(RollupShardIndexerStatus.COMPLETED), e -> {
+                logger.error("error while updating downsampling persistent task state", e);
+                throw new ElasticsearchException("error while updating downsampling persistent task state", e);
+            })
+        );
 
         return new DownsampleIndexerAction.ShardDownsampleResponse(indexShard.shardId(), task.getNumIndexed());
     }
 
+    private Query createQuery() throws IOException {
+        if (this.state.started() && this.state.tsid() != null) {
+            final SearchExecutionContext searchExecutionContext = indexService.newSearchExecutionContext(
+                indexShard.shardId().id(),
+                0,
+                searcher,
+                System::currentTimeMillis,
+                null,
+                Collections.emptyMap()
+            );
+            return new BoolQueryBuilder().filter(new TermQueryBuilder(TimeSeriesIdFieldMapper.NAME, this.state.tsid()))
+                .toQuery(searchExecutionContext);
+        } else if (this.state.done()) {
```

Comment:
maybe return in the `execute()` instead of here? This we can check without running query.